src/controllers/krate/downloads.rs

```diff
@@ -67,3 +67,70 @@ pub fn downloads(req: &mut dyn Request) -> AppResult<Response> {
         meta,
     }))
 }
+
+/// Handles the `GET /crates/:crate_id/recent_downloads` route.
+pub fn recent_downloads(req: &mut dyn Request) -> AppResult<Response> {
+    use diesel::dsl::*;
+
+    let ndays = 90;
+
+    let crate_name = &req.params()["crate_id"];
+    let conn = req.db_conn()?;
+    let krate = Crate::by_name(crate_name).first::<Crate>(&*conn)?;
+
+    // Get the versions for this crate
+    let versions = krate.all_versions().load::<Version>(&*conn)?;
+
+    #[derive(Debug, Serialize)]
+    struct Download<'a> {
+        version: &'a semver::Version,
+        downloads: i32,
+    }
+
+    #[derive(Debug, Serialize)]
+    struct Response<'a> {
+        downloads: Vec<Download<'a>>,
+        meta: Meta<'a>,
+    }
+
+    #[derive(Debug, Serialize)]
+    struct Meta<'a> {
+        #[serde(rename = "crate")]
+        krate: &'a str,
+        ndays: i32,
+    }
+
+    // Now get the grouped versions for the last `ndays` days.
+    //
+    // XXX I am not sure how to do this in the database yet, with Diesel's API, so for the time
+    // being, perform this aggregation in Rust.
+    let downloads = VersionDownload::belonging_to(versions.as_slice())
+        .filter(version_downloads::date.gt(date(now - ndays.days())))
+        .load(&*conn)?
+        .grouped_by(versions.as_slice())
+        .into_iter()
+        .map(|grouped_versions: Vec<VersionDownload>| {
+            let total_downloads = grouped_versions.iter().map(|v| v.downloads).sum();
+            // XXX(perf) this is slow, iterating over the versions array looking for the matching
```

Comment:
I think you can zip the results of the query with `versions` to get the matching version for each group, like [the search query does with crates and versions and other items here](https://github.com/rust-lang/crates.io/blob/3e5dd28127284e64ff8e4b106ca5dc3472398e5a/src/controllers/krate/search.rs#L177-L192)? There's some documentation with more explanation [here](https://docs.diesel.rs/diesel/associations/index.html) (search for `grouped_by` to get to the relevant section).

Reply:
This has been taken care of with the latest changes switching over to using diesel queries.